src/lib.rs

```diff
@@ -1,9 +1,361 @@
+    /// stream oriented transports
+    #[cfg(feature = "tokio")]
+    pub async fn connect_tokio_io<Io, Source>(
+        self,
+        source: Source,
+        io: Io,
+        direction: ConnDirection,
+    ) -> Result<(), error::RunConnection<error::CodecError, error::CodecError>>
+    where
+        Io: tokio::io::AsyncRead + tokio::io::AsyncWrite,
+    {
+        todo!()
+    }
+
+    /// Attempt to load the document given by `id` from storage
+    ///
+    /// If the document is not found in storage then `Ok(None)` is returned.
+    ///
+    /// Note that this _does not_ attempt to fetch the document from the
+    /// network.
+    pub async fn load(&self, id: DocumentId) -> Result<Option<DocHandle>, Store::Error> {
+        todo!()
+    }
+
+    /// Request a document from a peer in the network
+    ///
+    /// This will connect to the peers specified by `peers` and run the sync
+    /// protocol for the given document ID. Once the sync protocol has run (or
+    /// not as the case may be) this function will return the document handle.
+    ///
+    /// This can be used both to fetch a document you don't have and to request
+    /// updates you think another peer has but which you haven't been sent yet.
+    ///
+    /// # Arguments
+    ///
+    /// * id - The ID of the document to request
+    /// * peer - The peer(s) to request the document from.
+    pub async fn request(
+        &self,
+        id: DocumentId,
+        peers: PeerSpec,
+    ) -> Result<Option<DocHandle>, error::Request> {
+        todo!()
+    }
+
+    /// Create a new document, store it, and return a `DocHandle` for it
+    ///
+    /// In addition to creating the document this will also enqueue sync messages
+    /// to every connected peer for whom `Share::should_share` returns `ShareDecision::Share`
+    ///
+    /// # Errors
+    ///
+    /// * If a document with this ID already exists in storage
+    pub async fn create(&self, doc: am::Automerge) -> Result<DocHandle, error::Create<Store>> {
+        todo!()
+    }
+
+    /// Announce that we have a given document and wait for the peers specified by `ack_by` to sync
+    /// with us
+    ///
+    /// # Errors
+    ///
+    /// * If the given document does not exist
+    /// * If there is no connection to the peers given in `ack_by`
+    pub async fn announce(&self, doc: DocumentId, ack_by: PeerSpec) -> Result<(), error::Announce> {
+        todo!()
+    }
+
+    /// Shut down the repository
+    ///
+    /// This will signal to the background task that it should shutdown and
+    /// then wait for it to complete
+    ///
+    /// If the background task has already shutdown this returns immediately. If
+    /// another task has already signalled the background task to shutdown then
+    /// this will wait for shutdown to complete.
+    pub async fn shutdown(self) -> Result<(), error::Shutdown> {
+        todo!()
+    }
```

Comment:
I think the repo handle should be kept unique so that there is only one shutdown message ever, at least initially to keep shutdown simple.

Reply:
The problem with keeping the repo handle unique is that now if you're using this from an async context you have to set up a bunch of boilerplate to serialize all calls to the repo through some kind of channel. I think instead we should manage all that complexity internally and just give the users a clonable repo handle.

Reply:
In the examples so far I put the handle in a `Arc<Mutex<Option<_>>>` to deal with it. But yes a clonable handle would be more convenient.